google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java

```diff
@@ -0,0 +1,121 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Records stats used in dynamic flow control, the decaying average of recorded latencies and the
+ * last timestamp when the thresholds in {@link FlowController} are updated.
+ */
+final class DynamicFlowControlStats {
+
+  private static final double DEFAULT_DECAY_CONSTANT = 0.015; // Biased to the past 5 minutes
+
+  private AtomicLong lastAdjustedTimestampMs;
+  private DecayingAverage meanLatency;
+
+  DynamicFlowControlStats() {
+    this(DEFAULT_DECAY_CONSTANT);
+  }
+
+  DynamicFlowControlStats(double decayConstant) {
+    this.lastAdjustedTimestampMs = new AtomicLong(0);
+    this.meanLatency = new DecayingAverage(decayConstant);
+  }
+
+  void updateLatency(long latency) {
+    updateLatency(latency, System.currentTimeMillis());
+  }
+
+  @VisibleForTesting
+  void updateLatency(long latency, long timestampMs) {
+    meanLatency.update(latency, timestampMs);
+  }
+
+  double getMeanLatency() {
+    return getMeanLatency(System.currentTimeMillis());
+  }
+
+  @VisibleForTesting
+  double getMeanLatency(long timestampMs) {
+    return meanLatency.getMean(timestampMs);
+  }
+
+  public long getLastAdjustedTimestampMs() {
+    return lastAdjustedTimestampMs.get();
+  }
+
+  boolean setLastAdjustedTimestampMs(long last, long now) {
+    if (lastAdjustedTimestampMs.compareAndSet(last, now)) {
+      return true;
+    }
+    return false;
+  }
+
+  private class DecayingAverage {
+    private double decayConstant;
+    private double mean;
+    private double weightedCount;
+    private AtomicLong lastUpdateTimeInSecond;
+
+    DecayingAverage(double decayConstant) {
+      this.decayConstant = decayConstant;
+      this.mean = 0.0;
+      this.weightedCount = 0.0;
+      this.lastUpdateTimeInSecond = new AtomicLong(0);
+    }
+
+    synchronized void update(long value, long timestampMs) {
+      long now = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      Preconditions.checkArgument(
+          now >= lastUpdateTimeInSecond.get(), "can't update an event in the past");
+      if (lastUpdateTimeInSecond.get() == 0) {
+        lastUpdateTimeInSecond.set(now);
+        mean = value;
+        weightedCount = 1;
+      } else {
+        long prev = lastUpdateTimeInSecond.getAndSet(now);
+        long elapsed = now - prev;
+        double alpha = getAlpha(elapsed);
+        // Exponential moving average = weightedSum / weightedCount, where
+        // weightedSum(n) = value + alpha * weightedSum(n - 1)
+        // weightedCount(n) = 1 + alpha * weightedCount(n - 1)
+        // Using weighted count in case the sum overflows
+        mean =
+            mean * ((weightedCount * alpha) / (weightedCount * alpha + 1))
+                + value / (weightedCount * alpha + 1);
+        weightedCount = weightedCount * alpha + 1;
+      }
+    }
+
+    double getMean(long timestampMs) {
+      long now = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      Preconditions.checkArgument(
```

Comment:
I think there is a race condition here. update() can interleave with getMean. i think you need to just clamped elapsed to be non-negative